# EDA Notebook

In [1]:
import retriving_reviews
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

### Importing as the Yelp DataSets

In [2]:
data_reviews = retriving_reviews.load_pickle('/Users/gmgtex/Desktop/Galvanize/Immersive/capstone/pkl_data/yelp_reviews.pkl')
data_business = retriving_reviews.load_pickle('/Users/gmgtex/Desktop/Galvanize/Immersive/capstone/pkl_data/yelp_business.pkl')
# data_tips = retriving_reviews.load_pickle('/Users/gmgtex/Desktop/Galvanize/Immersive/capstone/pkl_data/yelp_tips.pkl')
# data_user = retriving_reviews.load_pickle('/Users/gmgtex/Desktop/Galvanize/Immersive/capstone/pkl_data/yelp_user.pkl')
# data_checkin = retriving_reviews.load_pickle('/Users/gmgtex/Desktop/Galvanize/Immersive/capstone/pkl_data/yelp_checkin.pkl')

## REVIEW DATASET

In [3]:
print(data_reviews.describe())
print(data_reviews.info())
data_reviews.head()

               cool         funny         stars        useful
count  4.736897e+06  4.736897e+06  4.736897e+06  4.736897e+06
mean   5.096600e-01  4.029171e-01  3.724048e+00  9.882921e-01
std    1.960374e+00  1.721954e+00  1.421104e+00  2.600021e+00
min    0.000000e+00  0.000000e+00  1.000000e+00  0.000000e+00
25%    0.000000e+00  0.000000e+00  3.000000e+00  0.000000e+00
50%    0.000000e+00  0.000000e+00  4.000000e+00  0.000000e+00
75%    0.000000e+00  0.000000e+00  5.000000e+00  1.000000e+00
max    5.130000e+02  6.310000e+02  5.000000e+00  1.125000e+03
<class 'pandas.core.frame.DataFrame'>
Int64Index: 4736897 entries, 0 to 4736896
Data columns (total 9 columns):
business_id    object
cool           int64
date           datetime64[ns]
funny          int64
review_id      object
stars          int64
text           object
useful         int64
user_id        object
dtypes: datetime64[ns](1), int64(4), object(4)
memory usage: 361.4+ MB
None


business_id  cool       date  funny               review_id  \
0  uYHaNptLzDLoV_JZ_MuzUA     0 2016-07-12      0  VfBHSwC5Vz_pbFluy07i9Q   
1  uYHaNptLzDLoV_JZ_MuzUA     0 2016-10-02      0  3zRpneRKDsOPq92tq7ybAA   
2  uYHaNptLzDLoV_JZ_MuzUA     0 2015-09-17      0  ne5WhI1jUFOcRn-b-gAzHA   
3  uYHaNptLzDLoV_JZ_MuzUA     0 2016-08-21      0  llmdwOgDReucVoWEry61Lw   
4  uYHaNptLzDLoV_JZ_MuzUA     0 2013-11-20      0  DuffS87NaSMDmIfluvT83g   

   stars                                               text  useful  \
0      5  My girlfriend and I stayed here for 3 nights a...       0   
1      3  If you need an inexpensive place to stay for a...       0   
2      3  Mittlerweile gibt es in Edinburgh zwei Ableger...       0   
3      4  Location is everything and this hotel has it! ...       0   
4      5  gute lage im stadtzentrum. shoppingmeile und s...       0   

                  user_id  
0  cjpdDjZyprfyDG3RlkVG3w  
1  bjTcT8Ty4cJZhEOEo01FGA  
2  AXgRULmWcME7J6Ix3I--ww  
3  oU2SSOmsp_A8JYI7Z2JJ5w  
4  0xtbPEna2Kei11vsU-U2Mw

In [4]:
text = data_reviews.text #these are the reviews that I will be processing
text.head()

0    My girlfriend and I stayed here for 3 nights a...
1    If you need an inexpensive place to stay for a...
2    Mittlerweile gibt es in Edinburgh zwei Ableger...
3    Location is everything and this hotel has it! ...
4    gute lage im stadtzentrum. shoppingmeile und s...
Name: text, dtype: object

In [5]:
data_reviews.text.iloc[2] #some reviews not in english

"Mittlerweile gibt es in Edinburgh zwei Ableger der Motel-One-Kette - diese hier ist eher das schlechtere.\n\nWir hatten auf unserer Schottlandtour die Gelegenheit beide Ableger kennenzulernen, da wir beide Nächte in Edinburgh (am ersten sowie am letzten Tag) in einem der beiden Motel One's verbrachten. Von diesem hier waren wir nach dem überraschend positiven Erlebnis im Motel One an der Princes Street [1] etwas enttäuscht.\n\nDas Hotel ist wesentlich größer und extrem unübersichtlich. Da man auf der fünften Etage noch Teile eines weiteren Gebäude, das mittels einer Glasbrücke mit dem Hauptgebäude verbunden ist, hinzugenommen hat, kann die Suche nach dem Zimmer insbesondere auf dieser Etage zu einem Gang im Labyrinth ausarten. Habe ich so noch in keinem Hotel erlebt - uns begegneten immer wieder Menschen, die sich auf der Suche nach ihren Zimmern verirrt haben.\nUnser Zimmer war ähnlich aufgebaut wie das Zimmer im Motel One Princes Street - befand sich jedoch im Inneren des Gebäudes u

In [27]:
data_reviews['sentiment'] = data_reviews['stars'].apply(lambda x: 'negative'if x <3 else ('neutral' if x==3 else 'positive'))
print(data_reviews['text'][(data_reviews['sentiment'] == 'negative')])
data_reviews.head()

16         This place is horrible, we were so excited to ...
21         Decent customer service but the food was awful...
24         The staff here is great and they're nice,  won...
25         I had the garlic ginger broccoli chicken and i...
29         Definitely not a fan. Coming from Orange Count...
31         I wish I could give 1.5 stars. Nothing special...
32         Disappointed that on yelp their hours show the...
35         Typical big business chinese place. Slightly f...
42         If I could give no stars I would. My family pl...
46         I really don't understand how anyone can eat t...
47         Went in for lunch with a coworker today. The s...
49         OH MY GOD, this place. I used to like Pei Wei ...
51         Haven't eaten at this location in particular, ...
60         We went as a family of 6 and was treated very ...
62         We are winter residents and always look forwar...
68         Took the family there last week.  Ordered frie...
72         In Surprise t

business_id  cool       date  funny               review_id  \
0  uYHaNptLzDLoV_JZ_MuzUA     0 2016-07-12      0  VfBHSwC5Vz_pbFluy07i9Q   
1  uYHaNptLzDLoV_JZ_MuzUA     0 2016-10-02      0  3zRpneRKDsOPq92tq7ybAA   
2  uYHaNptLzDLoV_JZ_MuzUA     0 2015-09-17      0  ne5WhI1jUFOcRn-b-gAzHA   
3  uYHaNptLzDLoV_JZ_MuzUA     0 2016-08-21      0  llmdwOgDReucVoWEry61Lw   
4  uYHaNptLzDLoV_JZ_MuzUA     0 2013-11-20      0  DuffS87NaSMDmIfluvT83g   

   stars                                               text  useful  \
0      5  My girlfriend and I stayed here for 3 nights a...       0   
1      3  If you need an inexpensive place to stay for a...       0   
2      3  Mittlerweile gibt es in Edinburgh zwei Ableger...       0   
3      4  Location is everything and this hotel has it! ...       0   
4      5  gute lage im stadtzentrum. shoppingmeile und s...       0   

                  user_id sentiment  
0  cjpdDjZyprfyDG3RlkVG3w  positive  
1  bjTcT8Ty4cJZhEOEo01FGA   neutral  
2  AXgRULmWcME7J6Ix3I--ww   neutral  
3  oU2SSOmsp_A8JYI7Z2JJ5w  positive  
4  0xtbPEna2Kei11vsU-U2Mw  positive

In [21]:
print(data_reviews['stars'].value_counts())
data_reviews['stars'].mean() # mean review rating

5    1988003
4    1135830
1     639849
3     570819
2     402396
Name: stars, dtype: int64


3.724048253529684

## BUSINESS DATASET

In [6]:
print(data_business.describe())
print(data_business.info())
data_business.head()

             is_open       latitude      longitude   review_count  \
count  156639.000000  156638.000000  156638.000000  156639.000000   
mean        0.844375      38.585033     -92.856485      30.238159   
std         0.362501       5.399871      26.557741      96.486631   
min         0.000000     -36.086009    -142.466650       3.000000   
25%         1.000000      33.627161    -112.138207       4.000000   
50%         1.000000      36.142381     -89.523198       9.000000   
75%         1.000000      43.596845     -79.668760      23.000000   
max         1.000000      89.999314     115.086769    6979.000000   

               stars  
count  156639.000000  
mean        3.647154  
std         0.977640  
min         1.000000  
25%         3.000000  
50%         3.500000  
75%         4.500000  
max         5.000000  
<class 'pandas.core.frame.DataFrame'>
Int64Index: 156639 entries, 0 to 156638
Data columns (total 15 columns):
address         156639 non-null object
attributes      15663

address  \
0                      691 Richmond Rd   
1                       2824 Milton Rd   
2                  337 Danforth Avenue   
3  7702 E Doubletree Ranch Rd, Ste 300   
4                       4719 N 20Th St   

                                          attributes             business_id  \
0  {'RestaurantsPriceRange2': 2, 'BusinessParking...  YDf95gJZaq05wvo7hTQbbQ   
1  {'GoodForMeal': {'dessert': False, 'latenight'...  mLwM-h2YhXl2NCgdS84_Bw   
2  {'BusinessParking': {'garage': False, 'street'...  v2WhjAB3PIBA8J8VxG3wEg   
3                                                 {}  CVtCbSB1zUcUWg-9TNGTuQ   
4  {'GoodForMeal': {'dessert': False, 'latenight'...  duHFBe87uNSXImQmvBh87Q   

                                          categories              city  \
0                       [Shopping, Shopping Centers]  Richmond Heights   
1  [Food, Soul Food, Convenience Stores, Restaura...         Charlotte   
2                               [Food, Coffee & Tea]           Toronto   
3               [Professional Services, Matchmakers]        Scottsdale   
4                          [Sandwiches, Restaurants]           Phoenix   

                                               hours  is_open   latitude  \
0  {'Monday': '10:00-21:00', 'Tuesday': '10:00-21...        1  41.541716   
1  {'Monday': '10:00-22:00', 'Tuesday': '10:00-22...        0  35.236870   
2  {'Monday': '10:00-19:00', 'Tuesday': '10:00-19...        0  43.677126   
3  {'Tuesday': '9:00-17:00', 'Friday': '9:00-17:0...        1  33.565082   
4                                                 {}        0  33.505928   

    longitude                                name neighborhood postal_code  \
0  -81.493116                Richmond Town Square                    44143   
1  -80.741976  South Florida Style Chicken & Ribs     Eastland       28215   
2  -79.353285                    The Tea Emporium    Riverdale     M4K 1N7   
3 -111.916400                            TRUmatch                    85258   
4 -112.038847                             Blimpie                    85016   

   review_count  stars state  
0            17    2.0    OH  
1             4    4.5    NC  
2             7    4.5    ON  
3             3    3.0    AZ  
4            10    4.5    AZ

### Important summary stats
#### Business Rating
* mean rating=3.647154
* std=0.977640
* min rating=1
* max rating=5
    * 25%=3.0
    * 50%=3.5
    * 75%=4.5
        * break up sentiment anaylsis --> below 50% negative, above 50% positive

#### Number of Reviews per Business
* mean rating=30.238159
* std=96.486631   
* min=3.000000 
* max=6979.000000
    * 25%=4.000000   
    * 50%=9.000000   
    * 75%=23.000000  

In [7]:
data_business.city.value_counts() #unique locations and counts

Las Vegas                          24768
Phoenix                            15656
Toronto                            15483
Charlotte                           7557
Scottsdale                          7510
Pittsburgh                          5688
Montréal                            5175
Mesa                                5146
Henderson                           4130
Tempe                               3949
Chandler                            3649
Edinburgh                           3625
Cleveland                           2979
Madison                             2891
Glendale                            2841
Gilbert                             2835
Mississauga                         2316
Stuttgart                           1975
Peoria                              1527
Markham                             1390
North Las Vegas                     1254
Champaign                           1084
North York                           944
Surprise                             934
Scarborough     

In [8]:
cats = data_business.categories
print(cats.shape)
cats.head(10)

(156639,)


0                         [Shopping, Shopping Centers]
1    [Food, Soul Food, Convenience Stores, Restaura...
2                                 [Food, Coffee & Tea]
3                 [Professional Services, Matchmakers]
4                            [Sandwiches, Restaurants]
5                            [Shopping, Tobacco Shops]
6                    [Chiropractors, Health & Medical]
7    [Automotive, Oil Change Stations, Car Wash, Au...
8    [Jewelry Repair, Gold Buyers, Local Services, ...
9                     [Health & Medical, Optometrists]
Name: categories, dtype: object

Need to figure out which key terms to split by in categories possibly Restaurants

In [9]:
scategories=[cat for cat in cats]
keywords = ['Restaurants','Food','Bars']
ind = set(ind for ind, cat in enumerate(scategories) for word in cat if word in keywords)
len(ind)

67882

In [19]:
restaurant_ids = data_business['business_id'].iloc[list(ind)]
print(len(restaurant_ids))
restaurant_ids.head()

67882


131072    23gakZ_NlN3K0SqOBybwug
1         mLwM-h2YhXl2NCgdS84_Bw
2         v2WhjAB3PIBA8J8VxG3wEg
4         duHFBe87uNSXImQmvBh87Q
131077    JRX7Cugs8NxqdgLS7G5q3Q
Name: business_id, dtype: object

In [22]:
restaurant_df = data_reviews[data_reviews['business_id'].isin(restaurant_ids)]
print(restaurant_df.describe())
print(restaurant_df.info())
restaurant_df.head()

               cool         funny         stars        useful
count  3.306658e+06  3.306658e+06  3.306658e+06  3.306658e+06
mean   5.325368e-01  4.109463e-01  3.709718e+00  9.373561e-01
std    2.022801e+00  1.774972e+00  1.343857e+00  2.665687e+00
min    0.000000e+00  0.000000e+00  1.000000e+00  0.000000e+00
25%    0.000000e+00  0.000000e+00  3.000000e+00  0.000000e+00
50%    0.000000e+00  0.000000e+00  4.000000e+00  0.000000e+00
75%    0.000000e+00  0.000000e+00  5.000000e+00  1.000000e+00
max    5.130000e+02  6.310000e+02  5.000000e+00  1.125000e+03
<class 'pandas.core.frame.DataFrame'>
Int64Index: 3306658 entries, 16 to 4736896
Data columns (total 9 columns):
business_id    object
cool           int64
date           datetime64[ns]
funny          int64
review_id      object
stars          int64
text           object
useful         int64
user_id        object
dtypes: datetime64[ns](1), int64(4), object(4)
memory usage: 252.3+ MB
None


business_id  cool       date  funny               review_id  \
16  jQsNFOzDpxPmOurSWCg1vQ     0 2017-06-03      0  ByRzJ8rF2KJWLr-cUNU6EA   
17  jQsNFOzDpxPmOurSWCg1vQ     0 2015-03-26      0  i5UwUPlQFPLcE8p2gPFwBw   
18  jQsNFOzDpxPmOurSWCg1vQ     1 2012-12-30      1  EyQyvTTg2jX4or9bB8PC9g   
19  jQsNFOzDpxPmOurSWCg1vQ     1 2009-01-12      1  G-EFA005besj5uHsH0sQFA   
20  jQsNFOzDpxPmOurSWCg1vQ     0 2015-07-11      0  6PcJSGUBSLjt4VLXos5C4A   

    stars                                               text  useful  \
16      1  This place is horrible, we were so excited to ...       0   
17      4  For being fairly "fast" food.. Pei Wei (pronou...       1   
18      5  I decided to try it out, I'm celiac and theref...       2   
19      3  I'm not saying Pei Wei is the best asian food ...       1   
20      3  Sometimes the food is spot on and delicious an...       0   

                   user_id  
16  kzyLOqiJvyw_FWFTw2rjiQ  
17  WZXp9-V2dqRRJqhGgRqueA  
18  XylT12exfdLiI_3uDLVIpw  
19  Ji9PeffxjwqPLO7pEfSpKQ  
20  TLIWzAJPrET0zX4_vgvLhg

In [ ]:
text, review_rating = [], []
for i in range(restaurant_df.shape[0]):
    text.append(restaurant_df['text'].iloc[i])
    review_rating.append(restaurant_df['stars'].iloc[i])
print(len(text), len(review_rating))
text[1], review_rating[1]

## TIPS DATASET

In [23]:
print(data_tips.describe())
print(data_tips.info())
data_tips.head()

              likes
count  1.028802e+06
mean   1.674472e-02
std    1.469311e-01
min    0.000000e+00
25%    0.000000e+00
50%    0.000000e+00
75%    0.000000e+00
max    1.500000e+01
<class 'pandas.core.frame.DataFrame'>
Int64Index: 1028802 entries, 0 to 1028801
Data columns (total 5 columns):
business_id    1028802 non-null object
date           1028802 non-null datetime64[ns]
likes          1028802 non-null int64
text           1028802 non-null object
user_id        1028802 non-null object
dtypes: datetime64[ns](1), int64(1), object(3)
memory usage: 47.1+ MB
None


business_id       date  likes  \
0  tJRDll5yqpZwehenzE2cSg 2012-07-15      0   
1  jH19V2I9fIslnNhDzPmdkA 2015-08-12      0   
2  dAa0hB2yrnHzVmsCkN4YvQ 2014-06-20      0   
3  dAa0hB2yrnHzVmsCkN4YvQ 2016-10-12      0   
4  ESzO3Av0b1_TzKOiqzbQYQ 2017-01-28      0   

                                                text                 user_id  
0              Get here early enough to have dinner.  zcTZk7OG8ovAmh_fenH21g  
1  Great breakfast large portions and friendly wa...  ZcLKXikTHYOnYt5VYRO5sg  
2  Nice place. Great staff.  A fixture in the tow...  oaYhjqBbh18ZhU0bpyzSuw  
3                     Happy hour 5-7 Monday - Friday  ulQ8Nyj7jCUR8M83SUMoRQ  
4  Parking is a premium, keep circling, you will ...  ulQ8Nyj7jCUR8M83SUMoRQ

In [29]:
tips = data_tips.text 
tips.iloc[1]

'Great breakfast large portions and friendly waitress. I highly recommend it'

Tips could be looked at as shorter review like objects?

## USER DATASET

In [30]:
print(data_user.describe())
print(data_user.info())
data_user.head()

       average_stars  compliment_cool  compliment_cute  compliment_funny  \
count   1.183362e+06     1.183362e+06     1.183362e+06      1.183362e+06   
mean    3.711835e+00     3.761705e+00     2.551341e-01      3.761705e+00   
std     1.105231e+00     9.143469e+01     1.482537e+01      9.143469e+01   
min     1.000000e+00     0.000000e+00     0.000000e+00      0.000000e+00   
25%     3.130000e+00     0.000000e+00     0.000000e+00      0.000000e+00   
50%     3.890000e+00     0.000000e+00     0.000000e+00      0.000000e+00   
75%     4.570000e+00     0.000000e+00     0.000000e+00      0.000000e+00   
max     5.000000e+00     2.959700e+04     1.361600e+04      2.959700e+04   

       compliment_hot  compliment_list  compliment_more  compliment_note  \
count    1.183362e+06     1.183362e+06     1.183362e+06     1.183362e+06   
mean     2.749478e+00     1.122869e-01     4.098577e-01     1.798434e+00   
std      7.917064e+01     1.272235e+01     1.556566e+01     6.027984e+01   
min      0.

average_stars  compliment_cool  compliment_cute  compliment_funny  \
0           3.80             5174              284              5174   
1           3.94             1556              211              1556   
2           4.72               15                1                15   
3           3.76                9                0                 9   
4           4.23              276                0               276   

   compliment_hot  compliment_list  compliment_more  compliment_note  \
0            5175               78              299             1435   
1            1285              101              134             1295   
2               5                0                1               11   
3               1                0                1                4   
4              59                0                8               51   

   compliment_photos  compliment_plain      ...         cool  \
0               7829              7397      ...        16856   
1                162              2134      ...        40110   
2                  8                20      ...           55   
3                  1                11      ...            4   
4                169               386      ...         6006   

                                               elite  fans  \
0   [2014, 2016, 2013, 2011, 2012, 2015, 2010, 2017]   209   
1  [2014, 2017, 2011, 2012, 2015, 2009, 2013, 200...   835   
2                                                 []    17   
3                                       [2016, 2017]    11   
4                                       [2017, 2016]    49   

                                             friends  funny    name  \
0  [M19NwFwAXKRZzt8koF11hQ, QRcMZ8pJJBBZaKubHOoMD...  16605     Cin   
1  [eoSSJzdprj3jxXyi94vDXg, QF0urZa-0bxga17ZeY-9l...  10882  Andrea   
2  [Oa84FFGBw1axX8O6uDkmqg, SRcWERSl4rhm-Bz9zN_J8...      4    Gabe   
3  [96DJovjKAtExnyBZRfniPQ, Djczq-bRn761U4FW77NUt...      4    Leah   
4  [iN0A6QIrEFYoSGHFaknh8Q, B2HDoWNIzLlon0IhS1cmD...    360    Juan   

   review_count useful                 user_id  yelping_since  
0           272  17019  lsSiIjAKVl-QRxKjRErBeg     2010-07-13  
1          2559  83681  om5ZiponkpRqUNa3pVPiRg     2006-01-18  
2           277     45  -lGwMGHMC_XihFJNKCJNRg     2014-10-31  
3           436     15  D-ydMTPGWXTVm4_jjp0k9g     2013-04-01  
4           921   9152  PcvbBOCOcs6_suRDH7TSTg     2012-08-16  

[5 rows x 22 columns]

## Shower Thoughts on User Dataset
* looks like users get a rating also --> higher rating could mean more reliable
* Does elite mean trustworthy or just someone that does a lot or reviews?
    * FROM YELP WEBSITE:
        * "What does it take to be Elite? Elite-worthiness is based on well-written reviews, high quality photos, an active voting and complimenting record, and a history of playing well with others."
        * You have to apply to be Elite
* useful count maybe an indicator of real review vs. fake/spam review.

In [ ]:
print(data_checkin.describe())
print(data_checkin.info())
data_checkin.head()

In [ ]:
sort_reviews_by_business = data_reviews[['business_id','text']].groupby('business_id').count().sort_values('text', ascending=False)
num_reviews_per_business = [['name','review_count','stars']].sort_values('review_count', ascending=False)
num_reviews_per_w_bus_id = data_business[['name', 'business_id' ,'review_count','stars']].sort_values('review_count', ascending=False)